File: Java/Shuffle.java
// ©2023 JDSherbert. All rights reserved.

import java.util.Random;

// Utility class for shuffling arrays using various algorithms.
public class Shuffle 
{
    // Fisher-Yates Shuffle Algorithm for shuffling elements in an array.
    // This algorithm guarantees a uniform random permutation.
    // Returns a copy of the shuffled array.
    // For more information, see https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle"
    public static <T> T[] fisherYatesShuffle(T[] array) 
    {
        Random rand = new Random();

        for (int n = array.length - 1; n > 0; n--)
        {
            int k = rand.nextInt(n + 1);

            // Swap the elements at indices 'n' and 'k'
            T temp = array[n];
            array[n] = array[k];
            array[k] = temp;
        }
    }
}

File: Python/Shuffle.py
# ©2023 JDSherbert. All rights reserved.
import random

# Utility class for shuffling arrays using various algorithms.
class Shuffle:
    @staticmethod
    def FisherYates(array):
        # Clone the array to avoid modifying the original
        shuffledArray = array.copy()

        # Iterate through the array in reverse order
        for n in range(len(shuffled_array) - 1, 0, -1):
            # Generate a random index 'k' between 0 and n (inclusive)
            k = random.randint(0, n)

            # Swap (using tuple deconstruction) the elements at indices 'k' and 'n'
            shuffled_array[n], shuffled_array[k] = shuffled_array[k], shuffled_array[n]

        # Return the shuffled array
        return shuffled_array
